activerecord/lib/active_record/future_result.rb

```diff
@@ -0,0 +1,96 @@
+# frozen_string_literal: true
+
+module ActiveRecord
+  class FutureResult
+    Canceled = Class.new(ActiveRecordError)
+
+    delegate :empty?, :to_a, to: :result!
+
+    def initialize(pool, *args, **kwargs)
+      @mutex = Mutex.new
+
+      @pool = pool
+      @args = args
+      @kwargs = kwargs
+
+      @executed = false
+      @error = nil
+      @result = nil
+    end
+
+    def schedule!
+      ActiveRecord::Base.asynchronous_queries_tracker.register(self)
```

Comment:
Aaron suggested passing in the tracker to `initialize` and storing it in an ivar.